feat: OLE CF and VBA modules implemented

davidmagnotti opened the pull request:
OLE CF (Object Linking and Embedding Compound File) format is a file format used for legacy Microsoft Office files, such as documents, workbooks, presentations, and others. It's also used with Visual Basic for Applications (VBA) which is known more commonly as Office macros.

I've implemented two modules for parsing OLE CF files and VBA. I've also expanded the `dump` command to allow dumping of file metadata such as stream metadata (OLE CF) and macros (VBA).

An example of how you could use this to identify the use of auto-execute macro method names like "Document_New":
```
import "vba"

rule detect_document_new
{
    condition:
        for any module in vba.module_code : (
            module matches /document_new/i
        )
}
```

davidmagnotti committed:
feat: OLE CF and VBA Modules Added

- Added support for parsing OLE CF and VBA (macro-enabled Office) files.

davidmagnotti committed:
Fix for OOM from Infinite Loop

- Addressed infinite loop issue in OLE CF parser.

davidmagnotti committed:
Merge branch 'main' of https://github.com/davidmagnotti/yara-x

plusvic committed:
fuzz: implement fuzzer for `vba` module.

plusvic committed:
style: fix clippy warnings.

plusvic committed:
style: fix clippy warning

davidmagnotti commented on lib/src/modules/olecf/parser.rs:
```diff
@@ -0,0 +1,400 @@
+                break;
+            }
+            current = next;
+        }
```
Great catch, fixed with this change that's committed (confirmed it correctly processes the attached file, whereas previously it didn't):
```
        while current < MAX_REGULAR_SECTOR {
            // Prevent cycles by keeping track of visited sectors
            if chain.contains(&current) {
                // We've seen this sector before - it's a cycle
                break;
            }
            
            chain.push(current);
            
            let next = match self.get_fat_entry(current) {
                Ok(n) => n,
                Err(_) => break,
            };
            
            // Check validity of next sector
            if next >= MAX_REGULAR_SECTOR || next == FREESECT || next == ENDOFCHAIN {
                break;
            }
            
            current = next;
        }
```

Thank you for addressing the other changes, too.

plusvic committed:
chore: remove `println`

plusvic committed:
style: apply `rustfmt`

plusvic commented on lib/src/modules/olecf/parser.rs:
```diff
@@ -0,0 +1,407 @@
+    }
+
+    fn get_regular_stream_data(&self, start_sector: u32, size: u64) -> Result<Vec<u8>, &'static str> {
+        let mut data = Vec::with_capacity(size as usize);
```
I highly recommend you to run the fuzzer in your side for a while to find issues in the parser. The fuzzer usually does a great a job at finding bugs. You can follow these steps:

```
rustup default nightly
cd lib/fuzz
cargo fuzz run vba_parser
```

plusvic committed:
style: apply `rustfmt`

plusvic committed:
refactor: some changes to make `parse_header` easier to follow.

plusvic committed:
Merge branch 'main' into vba

plusvic committed:
refactor: simplify the parsing of the DIFAT entries after the header.

plusvic commented:
According to the specification the first 109 DIFAT entries are located after the header, in the first sector, but the DIFAT can contain more than 109 entries, and I can't see where the possibly remaining entries are parsed. It looks like only the first 109 entries are parsed. This could be that a FAT larger than 109 sectors won't be parsed correctly.

plusvic committed:
refactor: simplify `follow_chain` function.

plusvic committed:
refactor: put stream names and sizes under a structure that represents a stream.

davidmagnotti committed:
Fixed OOM found from fuzzer

davidmagnotti committed:
Merge remote-tracking branch 'upstream/main'

plusvic committed:
Merge branch 'main' into ole

plusvic committed:
fix: enhance OLE CF parser with DIFAT chain and UTF-16 name support

Correctly decodes OLECF directory entry names from UTF-16 little-endian to prevent misinterpretation. Implements full traversal of the DIFAT chain to support larger OLECF files that require more FAT sectors than can be stored in the header. Adds cycle detection for MiniFAT chains to improve robustness against malformed input.

plusvic committed:
refactor: centralize OLE parsing for VBA project extraction

Extracts common OLE parsing logic into a shared function to reduce duplication and improve reusability across different extraction paths (e.g., direct OLE and OLE embedded in ZIP).
